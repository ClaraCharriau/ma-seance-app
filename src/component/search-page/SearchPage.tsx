import { useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';
import { searchMovies } from '../../client/movies/movies.client';
import { searchTheaters } from '../../client/theaters/theaters.client';
import { Movie } from '../../models/Movie';
import { Theater } from '../../models/Theater';
import MovieList from '../common/movie-list/MovieList';
import Spinner from '../common/spinner/Spinner';
import TheatersList from '../common/theaters-list/TheatersList';
import style from './SearchPage.module.css';

export const SearchPage = () => {
    const { search, state } = useLocation();
    const [movieResults, setMovieResults] = useState<Movie[]>([]);
    const [theaterResults, setTheaterResults] = useState<Theater[]>([]);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState('');
    const [isMovieSearch, setIsMovieSearch] = useState<boolean>(false);

    const ERROR_MESSAGE = "Une erreur s'est produite lors de votre recherche. Veuillez réessayer.";

    useEffect(() => {
        setError('');
        setLoading(true);
        const timeout = setTimeout(() => {
            setError(ERROR_MESSAGE);
            setLoading(false);
        }, 5000); // 5s

        if (!isMovieSearch) {
            searchTheaters(state.query)
                .then(response => {
                    setTheaterResults(response);
                    setLoading(false);
                    clearTimeout(timeout);
                })
                .catch(() => {
                    setError(ERROR_MESSAGE);
                    setLoading(false);
                });
        } else {
            searchMovies(state.query)
                .then(response => {
                    setMovieResults(response);
                    setLoading(false);
                    clearTimeout(timeout);
                })
                .catch(() => {
                    setError(ERROR_MESSAGE);
                    setLoading(false);
                });
        }

        return () => clearTimeout(timeout);
    }, [state.query, isMovieSearch]);

    const handleChange = () => {
        setIsMovieSearch(!isMovieSearch);
    };

    return (
        <main>
            <p>Résultats de la recherche "{state.query ?? search.split('?q=')}"</p>

            <div>
                <label className={style.switch}>
                    <input
                        className={style.switchInput}
                        type="checkbox"
                        checked={isMovieSearch}
                        onChange={handleChange}
                    />
                    <span className={style.switchSlider}></span>
                    <span className={style.cinemaLabel}>Cinémas</span>
                    <span className={style.filmLabel}>Films</span>
                </label>
            </div>

            {loading ? (
                <Spinner />
            ) : error ? (
                <div className={style.error}>{error}</div>
            ) : isMovieSearch && movieResults.length > 0 ? (
                <MovieList movieList={movieResults} />
            ) : !isMovieSearch && theaterResults.length > 0 ? (
                <TheatersList theaters={theaterResults} />
            ) : (
                <p>Aucun {isMovieSearch ? 'film' : 'cinéma'} ne corresponds à votre recherche.</p>
            )}
        </main>
    );
};
