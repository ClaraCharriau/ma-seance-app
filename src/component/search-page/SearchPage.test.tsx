import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { SearchPage } from './SearchPage';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useLocation: () => ({
        state: {
            query: 'test'
        }
    })
}));

describe('Search page tests', () => {
    it('should render search page', () => {
        // Given
        const component = render(
            <BrowserRouter>
                <SearchPage />
            </BrowserRouter>
        );

        // Then
        expect(component.baseElement).toMatchSnapshot();
    });
});
