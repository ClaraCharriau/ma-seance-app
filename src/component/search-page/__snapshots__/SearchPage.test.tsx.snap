// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search page tests should render search page 1`] = `
<body>
  <div>
    <main>
      <p>
        Résultats de la recherche "
        test
        "
      </p>
      <div>
        <label
          class="switch"
        >
          <input
            class="switchInput"
            type="checkbox"
          />
          <span
            class="switchSlider"
          />
          <span
            class="cinemaLabel"
          >
            Cinémas
          </span>
          <span
            class="filmLabel"
          >
            Films
          </span>
        </label>
      </div>
      <div
        class="spinner"
      />
    </main>
  </div>
</body>
`;
